python/pyspark/sql/session.py

```diff
@@ -446,33 +446,10 @@ def _createFromPandasWithArrow(self, pdf, schema):
             else:
                 schema = schema_from_arrow
         else:
-            batches = []
-            for i, pdf_slice in enumerate(pdf_slices):
-
-                # convert to series to pyarrow.Arrays to use mask when creating Arrow batches
-                arrs = []
-                names = []
-                for c, (_, series) in enumerate(pdf_slice.iteritems()):
-                    field = schema[c]
-                    names.append(field.name)
-                    t = to_arrow_type(field.dataType)
-                    try:
-                        # NOTE: casting is not necessary with Arrow >= 0.7
-                        arrs.append(pa.Array.from_pandas(_cast_pandas_series_type(series, t),
-                                                         mask=series.isnull(), type=t))
-                    except ValueError as e:
-                        warnings.warn("Arrow will not be used in createDataFrame: %s" % str(e))
-                        return None
-                batches.append(pa.RecordBatch.from_arrays(arrs, names))
-
-                # Verify schema of first batch, return None if not equal and fallback without Arrow
-                if i == 0:
-                    schema_from_arrow = from_arrow_schema(batches[i].schema)
-                    if schema != schema_from_arrow:
-                        warnings.warn("Arrow will not be used in createDataFrame.\n" +
-                                      "Supplied schema: %s\n!=\nArrow schema: %s"
-                                      % (str(schema), str(schema_from_arrow)))
-                        return None
+            arrow_types = [to_arrow_type(f.dataType) for f in schema.fields]
+            batches = [_create_batch([(c, t)
+                                      for (_, c), t in zip(pdf_slice.iteritems(), arrow_types)])
```

Comment:
BTW, looks this is only the place using `zip`. Not a big deal but I think we are safe to replace

```python
from itertools import imap as map
```

```python
from itertools import izip as zip, imap as map
```

We could change this after merging this first too.

Reply:
done